src/lib/car.js

```diff
@@ -0,0 +1,101 @@
+'use strict'
+
+const config = require('../config')
+const { elapsed } = require('./logging')
+const { openS3Stream } = require('./source')
+const { createDynamoItem, readDynamoItem } = require('./storage')
+const { publish } = require('./publish')
+const { now } = require('./util')
+const { storeBlocks, publishBlocks } = require('./block')
+
+function validateCar(id) {
+  try {
+    const info = id.match(/([^/]+)\/([^/]+)\/(.+)/)
+    if (!info) {
+      return
+    }
+    const [, bucketRegion, bucketName, key] = info
+    const url = new URL(`s3://${bucketName}/${key}`)
+    return { url, id, bucketRegion, bucketName, key }
+    /* c8 ignore next */
+  } catch (error) {}
+}
+
+/**
+ * Load the file from input
+ * @async
+ */
+function openSource({ car, logger }) {
+  return openS3Stream({ bucketRegion: car.bucketRegion, url: car.url, logger })
+}
+
+/**
+ * Load car from db
+ * @async
+ */
+function retrieveCar({ id, logger }) {
+  return readDynamoItem({ table: config.carsTable, keyName: config.carsTablePrimaryKey, keyValue: id, logger })
+}
+
+async function createCar({ car, source, logger }) {
+  await createDynamoItem({
+    table: config.carsTable,
+    keyName: config.carsTablePrimaryKey,
+    keyValue: car.id,
+    data: {
+      bucket: car.bucketName,
+      bucketRegion: car.bucketRegion,
+      key: car.key,
+      createdAt: now(),
+      roots: Array.from(new Set(source.roots.map(r => r.toString()))),
+      version: source.version,
+      fileSize: source.length
+    },
+    logger
+  })
+}
+
+async function publishCar({ car, logger, queue = config.notificationsQueue }) {
+  await publish({ queue, message: car.id, logger })
+}
+
+async function storeCar({ id, skipExists, logger }) {
+  const start = process.hrtime.bigint()
+
+  const car = validateCar(id)
+  if (!car) {
+    logger.error('Invalid CAR file format')
+    throw new Error('Invalid CAR file format')
+  }
+
+  if (skipExists) {
+    const exists = await retrieveCar({ id: car.id, logger })
+    if (exists) {
+      logger.info({ elapsed: elapsed(start) }, 'CAR already exists, skipping')
+      return
+    }
+  }
+
+  const source = await openSource({ car, logger })
+
+  await createCar({ car, source: source.indexer, logger })
+
+  /* c8 ignore next */
+  const onTaskComplete = config.skipPublishing ? undefined : publishBlocks
```

Comment:
There's nowhere we await on these calls to `publishBlocks` (the `queuedTasks` module calls `onTaskComplete` but does not await on it). What happens if the lambda handler function returns and there are async tasks in flight - does it wait for them to finish?

Reply:
How the lambda can finish before the tasks?
One of the possible issues is that there is an error on a concurrent set of batches, the tasks that are already fired have to finish, but I think it's ok because we are now getting a few failures (like 10 per week)